In [2]:
# Import Libraries

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

C:\Users\Nicholas\anaconda3\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\Nicholas\anaconda3\lib\site-packages\numpy\.libs\libopenblas.GK7GX5KEQ4F6UYO3P26ULGBQYHGQO7J4.gfortran-win_amd64.dll
C:\Users\Nicholas\anaconda3\lib\site-packages\numpy\.libs\libopenblas.WCDJNK7YVMPZQ2ME2ZZHJJRJ3JIKNDB7.gfortran-win_amd64.dll
C:\Users\Nicholas\anaconda3\lib\site-packages\numpy\.libs\libopenblas64__v0.3.21-gcc_10_3_0.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


In [6]:
# Load Data

GTD = pd.read_csv('C:/R Portfolio/Global_Terrorism_Prediction/globalterrorismdb_0522dist.csv', 
                     encoding = 'latin1',
                     low_memory = False)
GTD_1 = pd.read_csv('C:/R Portfolio/Global_Terrorism_Prediction/globalterrorismdb_2021Jan-June_1222dist.csv', 
                     encoding = 'latin1',
                     low_memory = False)
GTD_combined = pd.concat([GTD, GTD_1], ignore_index = True)
GTD_combined

eventid  iyear  imonth  iday  approxdate  extended  resolution  \
0       197000000001   1970       7     2         NaN         0         NaN   
1       197000000002   1970       0     0         NaN         0         NaN   
2       197001000001   1970       1     0         NaN         0         NaN   
3       197001000002   1970       1     0         NaN         0         NaN   
4       197001000003   1970       1     0         NaN         0         NaN   
...              ...    ...     ...   ...         ...       ...         ...   
214661  202106300023   2021       6    30         NaN         0         NaN   
214662  202106300029   2021       6    30  06/30/2021         0         NaN   
214663  202106300030   2021       6    30         NaN         1  08/07/2021   
214664  202106300038   2021       6    30         NaN         0         NaN   
214665  202107020037   2021       6    30         NaN         0         NaN   

        country         country_txt  region  ...  \
0            58  Dominican Republic       2  ...   
1           130              Mexico       1  ...   
2           160         Philippines       5  ...   
3            78              Greece       8  ...   
4           101               Japan       4  ...   
...         ...                 ...     ...  ...   
214661        4         Afghanistan       6  ...   
214662      138             Myanmar       5  ...   
214663      147             Nigeria      11  ...   
214664       45            Colombia       3  ...   
214665      138             Myanmar       5  ...   

                                                 addnotes  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
214661                                                NaN   
214662                                                NaN   
214663                                                NaN   
214664                                                NaN   
214665  There is doubt that this incident meets terror...   

                                                   scite1  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
214661  "Gunmen blow up power pylon in Parwan," Afghan...   
214662  "Spring Revolution Daily News for 16-30 June 2...   
214663  "Boko Haram Releases Abducted Catholic Priest ...   
214664  "Two dead and one wounded after clashes betwee...   
214665  "Spring Revolution Daily News for 2-7 July 202...   

                                                   scite2  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
214661  "Provinces hit by blackout after power pylon d...   
214662                                                NaN   
214663  "Kidnapped Maiduguri Catholic Priest regains f...   
214664                                                NaN   
214665                                                NaN   

                                                   scite3  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN

In [7]:
# Load Functions

execfile('C:/Python_Data_Sets/Functions 10_07_2023.py')

In [8]:
# Process Data

GTD_New = preprocess_data(GTD_combined)
GTD_New


Year  Month  Day       Country                       Region  \
0       1970      7    2  OtherCountry  Central America & Caribbean   
1       1970      0    0  OtherCountry                North America   
2       1970      1    0  OtherCountry               Southeast Asia   
3       1970      1    0  OtherCountry               Western Europe   
4       1970      1    0  OtherCountry                    East Asia   
...      ...    ...  ...           ...                          ...   
214661  2021      6   30   Afghanistan                   South Asia   
214662  2021      6   30  OtherCountry               Southeast Asia   
214663  2021      6   30  OtherCountry           Sub-Saharan Africa   
214664  2021      6   30  OtherCountry                South America   
214665  2021      6   30  OtherCountry               Southeast Asia   

             Province       City                Attack       Target  \
0       OtherProvince  OtherCity         Assassination      Private   
1       OtherProvince  OtherCity   HostageKidnapAttack  OtherTarget   
2       OtherProvince    Unknown         Assassination  OtherTarget   
3       OtherProvince  OtherCity            BombAttack  OtherTarget   
4       OtherProvince  OtherCity  InfrastructureAttack  OtherTarget   
...               ...        ...                   ...          ...   
214661  OtherProvince  OtherCity            BombAttack  OtherTarget   
214662  OtherProvince  OtherCity   HostageKidnapAttack  OtherTarget   
214663  OtherProvince    Unknown   HostageKidnapAttack  OtherTarget   
214664  OtherProvince    Unknown            BombAttack  OtherTarget   
214665  OtherProvince  OtherCity            BombAttack     Business   

             Group       Weapon  Dead  Lethal  
0       OtherGroup  OtherWeapon   1.0       1  
1       OtherGroup  OtherWeapon   0.0       0  
2       OtherGroup  OtherWeapon   1.0       1  
3       OtherGroup   Explosives   0.0       0  
4       OtherGroup   Incendiary   0.0       0  
...            ...          ...   ...     ...  
214661  OtherGroup   Explosives   0.0       0  
214662  OtherGroup  OtherWeapon   1.0       1  
214663  OtherGroup     Firearms   0.0       0  
214664  OtherGroup   Explosives   0.0       0  
214665  OtherGroup   Explosives   0.0       0  

[214666 rows x 13 columns]

In [24]:
# SA Region Data
SA_data = GTD_New[GTD_New['Region'] == 'South Asia']
SA_data

Year  Month  Day      Country      Region       Province       City  \
585     1970     11    1     Pakistan  South Asia  OtherProvince  OtherCity   
1186    1972      2   22        India  South Asia  OtherProvince  OtherCity   
1862    1973      5    1  Afghanistan  South Asia  OtherProvince  OtherCity   
2216    1974      2    2     Pakistan  South Asia  OtherProvince  OtherCity   
2704    1974     12    9     Pakistan  South Asia  OtherProvince  OtherCity   
...      ...    ...  ...          ...         ...            ...        ...   
214636  2021      6   29        India  South Asia  OtherProvince  OtherCity   
214648  2021      6   30  Afghanistan  South Asia  OtherProvince  OtherCity   
214649  2021      6   30  Afghanistan  South Asia  OtherProvince  OtherCity   
214650  2021      6   30     Pakistan  South Asia  OtherProvince  OtherCity   
214661  2021      6   30  Afghanistan  South Asia  OtherProvince  OtherCity   

                     Attack       Target       Group       Weapon  Dead  \
585           Assassination  OtherTarget  OtherGroup  OtherWeapon   4.0   
1186            OtherAttack  OtherTarget  OtherGroup   Explosives   0.0   
1862                Unknown  OtherTarget  OtherGroup  OtherWeapon   0.0   
2216             BombAttack  OtherTarget  OtherGroup     Firearms   0.0   
2704             BombAttack  OtherTarget  OtherGroup   Explosives   0.0   
...                     ...          ...         ...          ...   ...   
214636  HostageKidnapAttack      Private  OtherGroup     Firearms   1.0   
214648           BombAttack  OtherTarget     Taliban   Explosives   0.0   
214649           BombAttack      Private     Taliban   Explosives   2.0   
214650   ArmedAssaultAttack     Military  OtherGroup     Firearms   2.0   
214661           BombAttack  OtherTarget  OtherGroup   Explosives   0.0   

        Lethal  
585          1  
1186         0  
1862         0  
2216         0  
2704         0  
...        ...  
214636       1  
214648       0  
214649       1  
214650       1  
214661       0  

[56578 rows x 13 columns]

In [25]:
# Here we obtain the counts per variable for all variables

def get_value_counts(dataframe, variables):
    value_counts_dict = {}
    for variable in variables:
        if variable in dataframe.columns:
            value_counts_dict[variable] = dataframe[variable].value_counts()
        else:
            value_counts_dict[variable] = "Variable not found in the dataframe"
    return value_counts_dict

# List of variable names
variables_list = ['Country', 'Region', 'Province', 'City', 'Attack', 'Target', 'Group', 'Weapon', 'Dead','Lethal']

# Get value counts
result = get_value_counts(SA_data, variables_list)

# Print the value counts for each variable
for variable, value_counts in result.items():
    print(f"Value counts for {variable}:")
    print(value_counts)
    print()


Value counts for Country:
Afghanistan     20373
Pakistan        15654
India           14164
OtherCountry     6387
Name: Country, dtype: int64

Value counts for Region:
South Asia    56578
Name: Region, dtype: int64

Value counts for Province:
OtherProvince    56578
Name: Province, dtype: int64

Value counts for City:
OtherCity    54064
Unknown       2364
Name: City, dtype: int64

Value counts for Attack:
BombAttack              25530
ArmedAssaultAttack      13751
Assassination            5217
Unknown                  4557
HostageKidnapAttack      4184
InfrastructureAttack     2670
OtherAttack               669
Name: Attack, dtype: int64

Value counts for Target:
OtherTarget    13880
Private        12679
Police         11507
Military        8281
GovtGen         6650
Business        3581
Name: Target, dtype: int64

Value counts for Group:
OtherGroup    43644
Taliban       12934
Name: Group, dtype: int64

Value counts for Weapon:
Explosives     27388
Firearms       18242
OtherWeapon     8

In [26]:
observed = SA_data['Weapon'].value_counts()
observed

Explosives     27388
Firearms       18242
OtherWeapon     8161
Incendiary      2787
Name: Weapon, dtype: int64

For ease of illustration, as there are only four categories, we can explore weapon type. It appears that explosives is more 
likely to be chosen than incendiary weapon, but is this really true or is it down to chance? We can test it using Pearson’s 
goodness of fit test. Our research hypothesis is that weapon types are not chosen randomly.
Let us take as null hypothesis that there is equal probability of any given weapon type being used. 
There are four weapon types, so each type is given 25% probability. Our alternative hypothesis is that at least one of the 
weapon type choices probabilities isn’t .25

In [27]:
# make dictionary of values
SA_dict = {'Explosives' : .25,
           'Firearms' : .25,
           'OtherWeapon': .25,
           'Incendiary' : .25}
   
# create series from dictionary
probabilities = pd.Series(SA_dict)
probabilities

Explosives     0.25
Firearms       0.25
OtherWeapon    0.25
Incendiary     0.25
dtype: float64

# The “goodness of fit” test statistic

At this point, we have our observed frequencies and a collection of probabilities corresponding the null hypothesis that we 
want to test. We’ve stored these in Python as the corresponding variables observed and probabilities. What we want to do now is 
construct a test of the null hypothesis. We’re going to need a test statistic. 
The basic trick that a goodness of fit test uses is to construct a test statistic that measures how “close” the data are to 
the null hypothesis. If the data don’t resemble what you’d “expect” to see if the null hypothesis were true, then it probably
isn’t true. Okay, if the null hypothesis were true, what would we expect to see? Or, to use the correct terminology, what are 
the **expected frequencies*?. There are 56578 observations, and (if the null is true) the probability of any one of them 
choosing explosives weapon being chosen is 56578 * .25 = 14144.5, so I guess we’re expecting counts of explosives weapon, right? 
Lets calculate it:

In [28]:
# We have a sample size of 5342, so to calculate the expected weapon type use count is:

N = 56578 #sample size
expected = N * probabilities
expected

Explosives     14144.5
Firearms       14144.5
OtherWeapon    14144.5
Incendiary     14144.5
dtype: float64

Now, how do we translate this into a test statistic? Clearly, what we want to do is compare the expected number of observations 
in each category with the observed number of observations in that category. And on the basis of this comparison, we ought to be 
able to come up with a good test statistic. To start with, let’s calculate the difference between what the null hypothesis 
expected us to find and what we actually did find.

In [29]:
observed-expected

Explosives     13243.5
Firearms        4097.5
OtherWeapon    -5983.5
Incendiary    -11357.5
dtype: float64

Regardless of whether we do the calculations by hand or whether we do them in Python, it’s clear that more explosives were used
and fewer incendiary weapons than the null hypothesis predicted. However, a moment’s thought suggests that these raw differences
aren’t quite what we’re looking for. Intuitively, it feels like it’s just as bad when the null hypothesis predicts too few 
observations (which is what happened with explosives) as it is when it predicts too many (which is what happened with incendiary
weapons). So it’s a bit weird that we have a negative number for clubs and a positive number for hearts. One easy way to fix 
this is to square everything, so that we now calculate the squared differences. As before, we could do this by hand, but it’s 
easier to do it in Python…

In [30]:
(observed - expected)**2

Explosives     1.753903e+08
Firearms       1.678951e+07
OtherWeapon    3.580227e+07
Incendiary     1.289928e+08
dtype: float64

In [31]:
# If we convert these into actual numbers, they are as follows:
175,390,300
16,789,510
35,802,270
128,992,800
# Therefore, what we’ve got now is a collection of numbers that are significantly higher when the null hypothesis makes a bad 
#prediction (explosives and incendiary), but are small whenever it makes a good one (firearms and other weapon)

(128, 992, 800)

In [32]:
# Next, let’s also divide all these numbers by the expected frequency:
(observed - expected)**2 / expected

Explosives     12399.893404
Firearms        1186.998922
OtherWeapon     2531.179770
Incendiary      9119.644120
dtype: float64

In effect, what we’ve got here are four different “error” scores, each one telling us how big a “mistake” the null hypothesis 
made when we tried to use it to predict our observed frequencies. So, in order to convert this into a useful test statistic, 
one thing we could do is just add these numbers up. The result is called the goodness of fit statistic, conventionally referred 
to either as goodness of fitness test or GOF. We can calculate it using this command in Python:

In [33]:
sum((observed - expected)**2/expected)

25237.716214783133

Intuitively, it’s clear that if the gof is small, then the observed data are very close to what the null hypothesis predicted, 
so we’re going to need a large statistic in order to reject the null. As we’ve seen from our calculations, in our data set 
we’ve got a value of 25237.716214783133. So now the question becomes, is this a big enough value to reject the null?

The final step in the process of constructing our hypothesis test is to figure out what the rejection region is. That is, what 
values of the test statistic would lead us to reject the null hypothesis. As we saw earlier, large values imply that the null 
hypothesis has done a poor job of predicting the data from our experiment, whereas small values of imply that it’s actually 
done pretty well. Therefore, a pretty sensible strategy would be to say there is some critical value, such that if the value is 
bigger than the critical value we reject the null; but if the value is smaller than this value we retain the null. Right, so all 
we have to do is figure out what this critical value is. And it’s pretty straightforward. If we want our test to have 
significance level of 95% (that is, we are willing to tolerate a Type I error rate of 5%), then we have to choose our critical 
value so that there is only a 5% chance that the value could get to be that big if the null hypothesis is true. That is to say, 
we want the 95th percentile of the sampling distribution. The 95th percentile of a chi-squared distribution with 3 degrees of 
freedom can be calculated like this by using chi2.ppf() in python:

In [34]:
from scipy.stats import chi2
round(chi2.ppf(0.95, 3), 2)

7.81

Therefore, since our value is 25237.716214783133, we can reject the null hypothesis and conclude that weapon choices are not 
chosen randomly. If we want an exact value, we can calculate it using the chisquare() function from scipy.stats:

In [35]:
from scipy.stats import chisquare

chisquare(f_obs = observed, f_exp = expected)

Power_divergenceResult(statistic=25237.716214783133, pvalue=0.0)

A p-value of 0.0 suggests very strong evidence against the null hypothesis. 

In [36]:
# The above steps can also be done in this simpler code:
import pandas as pd
from scipy.stats import chisquare

ans = chisquare(SA_data['Weapon'].value_counts())

print("x2 = ", ans[0])
print("p = ", round(ans[1],3))

x2 =  25237.716214783133
p =  0.0


# The chi square test of association or independence

In [37]:
# First we subset the weapon type and group variables:
selected_columns = ['Group', 'Weapon']
SA_data_1 = SA_data[selected_columns]

In [45]:
observations = pd.crosstab(index = SA_data_1["Weapon"], columns = SA_data_1["Group"], margins = False)
observations

Group        OtherGroup  Taliban
Weapon                          
Explosives        22911     4477
Firearms          14323     3919
Incendiary         2487      300
OtherWeapon        3923     4238

In [41]:
# Our hypothesis is that groups attack using different weapons. 

In [54]:
# Now that we have a frequency table, we can use the chi2_contingency function from scipy.stats to crunch the numbers:

from scipy.stats import chi2_contingency

chi2, p, dof, ex = chi2_contingency(observations, correction = True) # Yates correction
print("chi2 = ", chi2)
print("p = ", p)
print("degrees of freedom = ", dof)
print("expected = ", ex)

chi2 =  4820.553450130939
p =  0.0
degrees of freedom =  3
expected =  [[21126.97288699  6261.02711301]
 [14071.79200396  4170.20799604]
 [ 2149.87853936   637.12146064]
 [ 6295.35656969  1865.64343031]]


In [ ]:
chi2 is the test statistic, p is the p-value, dof is the degrees of freedom for the test, and ex tells you the “expected”
values, that is, the frequency counts you would expect to see under the null hypothesis. 
The null hypothesis, by the way, is that the variables are independent of one another, that is, that the choice of weapon has 
nothing to do with whether the group is Taliban or OtherGroup. The alternative, of course, is that they are not independent;
that is, that each group will tend to make different choices. Or, put another way, that the frequency of each weapon type
chosen is contingent on whether or not the group is Taliban or OtherGroup.

In [46]:
conda install pingouin

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Nicholas\anaconda3

  added / updated specs:
    - pingouin


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    littleutils-0.2.2          |             py_0          10 KB  conda-forge
    mpmath-1.3.0               |     pyhd8ed1ab_0         428 KB  conda-forge
    outdated-0.2.2             |     pyhd8ed1ab_0          11 KB  conda-forge
    pandas-flavor-0.2.0        |             py_0           8 KB  conda-forge
    pingouin-0.5.3             |     pyhd8ed1ab_0         158 KB  conda-forge
    tabulate-0.9.0             |     pyhd8ed1ab_1          35 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         650 KB

The following NEW packages will be INSTALLED:

  littleutils        conda-forge/noarch::littleutils-0

In [53]:
import pingouin as pg

expected, observed, stats = pg.chi2_independence(SA_data, x = 'Weapon', y = 'Group', correction = True)# Yates correction

In [52]:
print("scipy's calculation of expected frequencies:")
print(" ")
chi2, p, dof, ex = chi2_contingency(observations)
print(ex)

print("-------")

print("pingouin's calculation of expected frequencies:")
print(" ")
expected, observed, stats = pg.chi2_independence(SA_data, x = 'Weapon', y = 'Group', correction = True)# Yates correction
print(expected)

scipy's calculation of expected frequencies:
 
[[21126.97288699  6261.02711301]
 [14071.79200396  4170.20799604]
 [ 2149.87853936   637.12146064]
 [ 6295.35656969  1865.64343031]]
-------
pingouin's calculation of expected frequencies:
 
Group          OtherGroup      Taliban
Weapon                                
Explosives   21126.972887  6261.027113
Firearms     14071.792004  4170.207996
Incendiary    2149.878539   637.121461
OtherWeapon   6295.356570  1865.643430


In [51]:
# The numbers are the same. Now let's do some statistical analysis:
expected, observed, stats = pg.chi2_independence(SA_data, x = 'Weapon', y = 'Group', correction = True) # Yates correction
stats

test    lambda         chi2  dof  pval    cramer  power
0             pearson  1.000000  4820.553450  3.0   0.0  0.291894    1.0
1        cressie-read  0.666667  4582.740065  3.0   0.0  0.284603    1.0
2      log-likelihood  0.000000  4247.705195  3.0   0.0  0.274002    1.0
3       freeman-tukey -0.500000  4098.099960  3.0   0.0  0.269133    1.0
4  mod-log-likelihood -1.000000  4021.389029  3.0   0.0  0.266603    1.0
5              neyman -2.000000  4057.511029  3.0   0.0  0.267797    1.0

In [ ]:
# With pval = 0.0, OtherGroup is more likely to prefer Explosives, Firearms and Incendiary, while the Taliban is more likely to
# favour otherweapon

# Effect Size

In [ ]:
So, let’s suppose that you’ve run your chi-square test, which turns out to be significant. So you now know that there is some 
association between your variables (independence test) or some deviation from the specified probabilities (goodness of fit 
test). Now you want to report a measure of effect size. That is, given that there is an association/deviation, how strong is 
it? There are several different measures that you can choose to report, and several different tools that you can use to 
calculate them. Pingouin provides you with one automatically: Cramer’s V. The idea here is that the cramers v statistic is 
    supposed to range between 0 (no at all association) and 1 (perfect association).

In [60]:
import pingouin as pg
expected, observed_pingouin, stats = pg.chi2_independence(SA_data, x = 'Weapon', y = 'Group')
stats

test    lambda         chi2  dof  pval    cramer  power
0             pearson  1.000000  4820.553450  3.0   0.0  0.291894    1.0
1        cressie-read  0.666667  4582.740065  3.0   0.0  0.284603    1.0
2      log-likelihood  0.000000  4247.705195  3.0   0.0  0.274002    1.0
3       freeman-tukey -0.500000  4098.099960  3.0   0.0  0.269133    1.0
4  mod-log-likelihood -1.000000  4021.389029  3.0   0.0  0.266603    1.0
5              neyman -2.000000  4057.511029  3.0   0.0  0.267797    1.0

In [ ]:
# The cramers v figures all are found between 0.20 and 0.30, so moderate effect size.

In [ ]:
# If the counts in the table are low - for example less than 5, it's a good idea to use Fisher's Exact test

In [55]:
import pandas as pd
from scipy.stats import fisher_exact

In [58]:
freq_table = pd.crosstab(index = SA_data["Group"], columns = SA_data["Lethal"],margins=False)
freq_table

Lethal          0      1
Group                   
OtherGroup  22989  20655
Taliban      2775  10159

In [59]:
oddsratio, pvalue = fisher_exact(freq_table)  

print("p = ", pvalue)

p =  0.0


In [ ]:
# with a p-value of 0.0, we can reject the null hypothesis that OtherGroup is just as likely to conduct a non-lethal attack as 
# the Taliban